src/Type/ComparisonHelper.php

```diff
@@ -0,0 +1,127 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type;
+
+use PHPStan\Php\PhpVersion;
+use PHPStan\ShouldNotHappenException;
+use PHPStan\Type\Constant\ConstantArrayType;
+use PHPStan\Type\Constant\ConstantBooleanType;
+use function array_keys;
+use function count;
+
+final class ComparisonHelper
+{
+
+	public static function looseCompareConstantScalars(ConstantScalarType $leftType, ConstantScalarType $rightType, PhpVersion $phpVersion): BooleanType
+	{
+		if ($phpVersion->castsNumbersToStringsOnLooseComparison()) {
+			$isNumber = new UnionType([
+				new IntegerType(),
+				new FloatType(),
+			]);
+
+			if ($leftType->isString()->yes() && $leftType->isNumericString()->no() && $isNumber->isSuperTypeOf($rightType)->yes()) {
+				$stringValue = (string) $rightType->getValue();
+				return new ConstantBooleanType($stringValue === $leftType->getValue());
+			}
+			if ($rightType->isString()->yes() && $rightType->isNumericString()->no() && $isNumber->isSuperTypeOf($leftType)->yes()) {
+				$stringValue = (string) $leftType->getValue();
+				return new ConstantBooleanType($stringValue === $rightType->getValue());
+			}
+		} else {
+			if ($leftType->isString()->yes() && $leftType->isNumericString()->no() && $rightType->isFloat()->yes()) {
+				$numericPart = (float) $leftType->getValue();
+				return new ConstantBooleanType($numericPart === $rightType->getValue());
+			}
+			if ($rightType->isString()->yes() && $rightType->isNumericString()->no() && $leftType->isFloat()->yes()) {
+				$numericPart = (float) $rightType->getValue();
+				return new ConstantBooleanType($numericPart === $leftType->getValue());
+			}
+			if ($leftType->isString()->yes() && $leftType->isNumericString()->no() && $rightType->isInteger()->yes()) {
+				$numericPart = (int) $leftType->getValue();
+				return new ConstantBooleanType($numericPart === $rightType->getValue());
+			}
+			if ($rightType->isString()->yes() && $rightType->isNumericString()->no() && $leftType->isInteger()->yes()) {
+				$numericPart = (int) $rightType->getValue();
+				return new ConstantBooleanType($numericPart === $leftType->getValue());
+			}
+		}
+
+		// @phpstan-ignore-next-line
+		return new ConstantBooleanType($leftType->getValue() == $rightType->getValue()); // phpcs:ignore
+	}
+
+	/**
+	 * @param callable(Type, Type): BooleanType $valueComparisonCallback
+	 */
+	public static function compareConstantArrayType(ConstantArrayType $leftType, ConstantArrayType $rightType, callable $valueComparisonCallback): BooleanType
```

Comment:
You've missed my point. This is supposed to be in `ConstantArrayType::looseCompare()`.

Reply:
So should I duplicate the logic?

Its also called from resolveIdentical.

Reply:
Or is the goal to also reduce resolveIdentical to just

isSuperTypeOf()->no() ?

Reply:
> Or is the goal to also reduce resolveIdentical to just

Yeah, probably, in the future. I'm fine if momentarily the code is a bit duplicated.

Reply:
just duplicated the method and removed the callable arg to improve readability